tests/test_emulator.cpp

```diff
@@ -168,33 +171,47 @@ class EmulatorTest : public testing::Test {
         em_ops.read_memory = test_read_memory;
         em_ops.write_memory = test_write_memory;
         em_ctxt.ops = &em_ops;
-        em_ctxt.mode = EM_MODE_PROT64;
         em_ctxt.vcpu = &vcpu;
         em_ctxt.rip = 0;
     }
 
-    void assemble_decode(const char* insn,
+    void assemble_decode(em_mode_t mode,
+                         const char* insn,
                          uint64_t rip,
                          size_t* size,
                          size_t* count,
                          em_status_t* decode_status) {
         uint8_t* code;
         int err;
 
-        err = ks_asm(ks, insn, 0, &code, size, count);
+        switch (mode) {
+        case EM_MODE_PROT16:
+            err = ks_asm(ks_x86_16, insn, 0, &code, size, count);
```

Comment:
nit: Instead of calling `ks_asm()` once per case, create a local `ks_engine*` and make it the focus of the switch-case.

Reply:
> Instead of calling `ks_asm()` once per case

But we need to assemble each test case, hence calling `ks_asm` once per case is unavoidable. Or am I misunderstanding something?

> defer the call until after the switch case by creating a local `ks_engine*`.

Wouldn't that hurt performance? That way we'd have to create, configure and destroy a new Keystone engine at every instruction we want to test, whereas currently we simply create 3 engines and switch between them based on the emulator mode.

Reply:
@raphaelning Do you mean this switch could just select a `ks_engine*`, to be used in one call to `ks_asm()` below the switch/case statement block?

Reply:
@PatrickvL Exactly, thanks for clarifying that for me. By "case" I mean a case in the switch-case statement, not a GTest test case... To be absolutely clear:

```c
        ks_engine* ks = nullptr;

        switch (mode) {
        case EM_MODE_PROT16:
            ks = ks_x86_16;
            break;
        // 3 more cases
        }
        ASSERT_EQ(KS_ERR_OK, ks_asm(ks_x86_16, insn, 0, &code, size, count));
```

Reply:
Surely you mean `ASSERT_EQ(KS_ERR_OK, ks_asm(ks, insn, 0, &code, size, count));`

Reply:
Oh, I fully agree with that. I've modified my last commit to include this change.

Now it's at 43d5fde.